File: java/minimumwindowsubstring.java
public class Solution {
    public String minWindow(String S, String T) {
        // Start typing your Java solution below
        // DO NOT write main() function
        int start = 0;
        int end = 0;
        int n =0;
        
        int min = Integer.MAX_VALUE;
        int minstart = 0;
        int minend = 0;
        HashSet<Character> hs = new HashSet<Character>();
        int count[] = new int[300];
        int sta[] = new int[300];
        for(int i=0;i<T.length();i++){
            hs.add(T.charAt(i));
            sta[T.charAt(i)+0]++;
        }
        int target = hs.size();
        boolean found = false;
        while(end<S.length()){
            while(n<target && end < S.length()){
                if(!hs.contains(S.charAt(end))){
                    end++;
                }else{
                    count[S.charAt(end)+0]++;
                    if(count[S.charAt(end)+0]==sta[S.charAt(end)+0]){
                        n++;
                    }
                    end++;
                }
            }
            if(min>end+1-start && n==target){
                found = true;
                min = end+1-start;
                minend = end;
                minstart = start;
            }
            
            while(n==target && start<=end){
                if(!hs.contains(S.charAt(start))){
                    start++;
                }else{
                    if(count[S.charAt(start)+0]==sta[S.charAt(start)+0]){
                        n--;
                    }
                    count[S.charAt(start)+0]--;
                    start++;
                }
            }
            if(min>end+2-start){
                min = end+2-start;
                minend = end;
                minstart = start-1;
            }
        }
        if(!found){
            return "";
        }
        return S.substring(minstart,minend);
    }
}

File: cpp/MinimumWindowSubstring.cpp
class Solution {
public:
    string minWindow(string S, string T) {
        int slen = (int)(S.length());
        int count[256];
        int countT[256];
        int totalCount = 0;
        int totalCountT = 0;
        for(int i=0;i<256;i++) {
            count[i] = 0;
            countT[i] = 0;
        }
        for(int i=0;i<(int)T.length();i++) {
            countT[T[i]]++;
            totalCountT++;
        }
        int start = 0;
        int end = 0;
        
        int minWindowLen = 0x7fffffff;
        string minWidowStr = "";
        
        while(end<slen) {
            if(countT[S[end]] > 0) {
                count[S[end]]++;
                if(count[S[end]] <= countT[S[end]]) {
                    totalCount++;
                }
                if(totalCount == totalCountT) {
                    while(start<=end) {
                        if(countT[S[start]] ==0 ) {
                            start++;
                        } else if(count[S[start]] > countT[S[start]]) {
                            count[S[start]]--;
                            start++;
                        } else {
                            break;
                        }
                    }
                    if(minWindowLen > end-start+1) {
                        minWindowLen = end-start+1;
                        minWidowStr = S.substr(start, minWindowLen);
                    }
                }
            }
            end++;
        }
        return minWidowStr;
    }
};
